menhir-recover/attributes.ml

```diff
@@ -229,44 +257,68 @@ struct
 
   let () =
     validate_attributes
-      ["recover.prelude"] "grammar attributes"
+      [ attr_names#prelude;
+        attr_names#default_cost_of_prod;
+        attr_names#default_cost_of_symbol;
+      ] "grammar attributes"
       Grammar.attributes;
     let symbol prj attrs =
       validate_attributes
-        ["recover.cost"; "recover.expr"] "symbol attributes"
-        (prj attrs)
+      [ attr_names#cost;
+        attr_names#expr;
+      ] "symbol attributes"
+      (prj attrs)
     in
     Nonterminal.iter (symbol G.Nonterminal.attributes);
     Terminal.iter (symbol G.Terminal.attributes);
     Production.iter (fun p ->
         validate_attributes
-          [ "recover.cost";
+          [ attr_names#cost;
             (* recover.expr: a lie to prevent warnings on an unfortunate
                interaction between menhir inlining and attributes *)
-            "recover.expr"
+            attr_names#expr;
           ] "production attributes"
         (Production.attributes p);
       Array.iter
         (fun (_,_,attrs) ->
-           validate_attributes ["recover.cost"] "item attributes" attrs)
+           validate_attributes [attr_names#cost] "item attributes" attrs)
         (Production.rhs p)
     )
 
-  let cost_of_attributes prj attrs =
-    List.fold_left (fun total attr ->
-      if Attribute.has_label "recover.cost" attr then
-        let payload = Attribute.payload attr in
-        let cost = if payload =  "inf"
-                   then Cost.infinite
-                   else Cost.of_int (int_of_string payload)
-        in Cost.add total cost
-      else total)
-      Cost.zero (prj attrs)
+  (* Read cost attributes returned by [prj attrs].
+     If there are many attributes with [name] return the sum of them. If there
+     are none of them return [default_cost]. *)
```

Comment:
`... there IS none ...`
Or maybe just `In opposite case...`

Reply:
fixed, thx